### PR TITLE
trace: simplify the Client interface; introduce StartEnd

### PR DESCRIPTION
**Pull request by rakyll**
I moved the responsibility of setting the casual relationship and start, end times to the finisher. Added StartEnd for those who want to create spans for already finished events.

### CODE REVIEW

File: trace.go [hunk 1]
```diff
@@ -28,9 +31,13 @@ type Client interface {
...
-	// Finish finishes the span in the context with the given labels. Nil labels
-	// should be accepted.
-	Finish(ctx context.Context, labels map[string][]byte) error
+	// Finish finishes the span in the context with the given labels.
+	//
+	// If casual is non-nil, span needs to be finalized with a casual relationship.
+	// Nil labels should be accepted.
+	//
+	// Span should have given start and end time.
+	Finish(ctx context.Context, causal []byte, labels map[string][]byte, start, end time.Time) error
```
Comment (bmizerany) [inline trace.go:40]: At this point I think some kind of `FinishParam` would be useful so we're not alway breaking the API if we want to add new args which is likely. We just want to be careful to not go crazy with it.

**Comment by bmizerany** [comment]
I had a quick comment, but LGTM.

File: trace.go (hunk 1 above)
Comment (rakyll) [reply trace.go:40]: I agree. We may even get rid of the StartEnd by allowing users to override start and end times at the finish time.

**Comment by rakyll** [comment]
Gonna merge this now, and will send a new PR with finish options.

**Pull request merged**